### ISSUE TITLE
Promise typings: returning unions in `then`

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#11835]
v2.0.3

``` ts
(() => {
    Promise.resolve(1).then((x): number | Promise<string> => {
        if (x) {
            return 2;
        } else {
            return Promise.resolve('foo')
        }
    })
})()
```

results in error:

```
[ts] Argument of type '(x: number) => number | Promise<string>' is not assignable to parameter of type '(value: number) => string | PromiseLike<string>'.
       Type 'number | Promise<string>' is not assignable to type 'string | PromiseLike<string>'.
         Type 'number' is not assignable to type 'string | PromiseLike<string>'.
```

Should this be possible? It is valid JS.

**Comment by 19majkel94** [comment]
Why not just `Promise<number|string>`?
Your API makes no sense - you have to remember what you put as a parameter and have `foo(x).then(y => ...)` or `y = foo(x)`.

**Comment by OliverJAsh** [comment]
It is quite conceivable to have a function that returns a union type of `A | Promise<B>`. I'm just wondering if we could increase the flexibility of the promise typings to allow for this usage.

**Comment by RyanCavanaugh** [comment]
If you have some improvement to the definition, feel free to send a PR.

**Comment by abgivant** [comment]
I just noticed a similar issue with returning union types. When the return type is a single promise type, everything works fine, but when returning a union of two different promise types, the resulting variable type ends up just being `any`.

![image](https://cloud.githubusercontent.com/assets/3431954/19898361/bf0acf14-a029-11e6-8288-ad0812d4227c.png)

However, the function signature does show that the unioned type is recognized.

![image](https://cloud.githubusercontent.com/assets/3431954/19898429/ee90eb6a-a029-11e6-801a-a433286f3b82.png)

And I tried the above suggestion of moving the union inside the type brackets, but that left the variable type as `any`, but also messed up the function signature.

![image](https://cloud.githubusercontent.com/assets/3431954/19898482/2ee14840-a02a-11e6-8496-bd402c44cd09.png)
